prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1121 @@
+	ofi_array_destroy(&srx->src_recv_queues);
+	ofi_array_destroy(&srx->src_trecv_queues);
+
+	while (!slist_empty(&srx->msg_queue)) {
+		entry = slist_remove_head(&srx->msg_queue);
+		(void) util_cancel_entry(srx, FI_SEND | FI_MSG,
+				container_of(entry, struct util_rx_entry,
+				             peer_entry));
+	}
+
+	while (!slist_empty(&srx->tag_queue)) {
+		entry = slist_remove_head(&srx->tag_queue);
+		(void) util_cancel_entry(srx, FI_SEND | FI_TAGGED,
+				container_of(entry, struct util_rx_entry,
+					     peer_entry));
+	}
+
+	while (!dlist_empty(&srx->all_unexp_msg)) {
+		dlist_pop_front(&srx->all_unexp_msg, struct util_rx_entry,
+				rx_entry, entry);
+		rx_entry->peer_entry.srx->peer_ops->discard_msg(
+							&rx_entry->peer_entry);
+		ofi_buf_free(rx_entry);
+	}
+
+	while (!dlist_empty(&srx->all_unexp_tag)) {
+		dlist_pop_front(&srx->all_unexp_tag, struct util_rx_entry,
+				rx_entry, entry);
+		rx_entry->peer_entry.srx->peer_ops->discard_tag(
+							&rx_entry->peer_entry);
+		ofi_buf_free(rx_entry);
+	}
+
+	ofi_array_destroy(&srx->src_unexp_msg_queues);
+	ofi_array_destroy(&srx->src_unexp_tag_queues);
+
+	ofi_atomic_dec32(&srx->cq->ref);
+	ofi_bufpool_destroy(srx->rx_pool);
+	ofi_mutex_destroy(&srx->lock);
+	free(srx);
+
+	return FI_SUCCESS;
+}
+
+static struct fi_ops util_srx_fid_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = util_srx_close,
+	.bind = util_srx_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static bool util_cancel_recv(struct util_srx_ctx *srx, struct slist *queue,
+			     uint64_t flags, void *context)
+{
+	struct slist_entry *item, *prev;
+	struct util_rx_entry *rx_entry;
+
+	srx->lock_acquire(&srx->lock);
+	slist_foreach(queue, item, prev) {
+		rx_entry = container_of(item, struct util_rx_entry, peer_entry);
+		if (rx_entry->peer_entry.context == context) {
+			slist_remove(queue, item, prev);
+			util_cancel_entry(srx, flags, rx_entry);
+			srx->lock_release(&srx->lock);
+			return true;
+		}
+	}
+	srx->lock_release(&srx->lock);
+	return false;
+}
+
+static int util_cancel_src(struct ofi_dyn_arr *arr, void *list, void *context)
+{
+	struct util_srx_ctx *srx;
+	struct slist *queue = list;
+	uint64_t flags;
+
+	srx = container_of(arr, struct util_srx_ctx, src_trecv_queues);
+
+	flags = arr == &srx->src_trecv_queues ?
+		FI_TAGGED | FI_RECV : FI_MSG | FI_RECV;
+
+	return (int) util_cancel_recv(srx, queue, flags, context);
+}
+
+static ssize_t util_srx_cancel(fid_t ep_fid, void *context)
+{
+	struct util_srx_ctx *srx;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	if (util_cancel_recv(srx, &srx->tag_queue, FI_TAGGED | FI_RECV,
+			     context))
+		return 0;
+
+	if (util_cancel_recv(srx, &srx->msg_queue, FI_MSG | FI_RECV,
+			     context))
+		return 0;
+
+	if (ofi_array_iter(&srx->src_trecv_queues, context, util_cancel_src))
+		return 0;
+
+	if (ofi_array_iter(&srx->src_recv_queues, context, util_cancel_src))
+		return 0;
+
+	return 0;
```

Comment:
Yeah the API is confusing to me for the cancel call - return 0 just means it was successfully submitted (whether it was found or not). The array iter return means it was found in that list so we can stop searching the rest of the lists. I'm updating it to FI_SUCCESS though instead of 0